src/hu_stuff.c

```diff
@@ -1153,21 +1153,42 @@ static void HU_widget_build_sttime(void)
   int offset = 0;
   extern int time_scale;
 
-  if (time_scale != 100)
+  if (gameaction == ga_completed)
+    plr->btuse_tics = 0;
+
+  if (hud_level_time & HUD_WIDGET_HUD ||
+      (hud_level_time & HUD_WIDGET_AUTOMAP && automapactive))
   {
-    offset += sprintf(hud_timestr, "\x1b%c%d%% ",
-            '0'+CR_BLUE, time_scale);
+    if (time_scale != 100)
+    {
+      offset += sprintf(hud_timestr, "\x1b%c%d%% ",
+              '0'+CR_BLUE, time_scale);
+    }
+
+    if (totalleveltimes)
+    {
+      const int time = (totalleveltimes + leveltime) / TICRATE;
+
+      offset += sprintf(hud_timestr + offset, "\x1b%c%d:%02d ",
+              '0'+CR_GREEN, time/60, time%60);
+    }
+
+    offset += sprintf(hud_timestr + offset, "\x1b%c%d:%05.2f",
+                      '0'+CR_GRAY, leveltime/TICRATE/60, (float)(leveltime%(60*TICRATE))/TICRATE);
+
+    if (plr->btuse_tics)
+      offset += sprintf(hud_timestr + offset, " ");
```

Comment:
You could just add the dummy space to the end of the regular format string some lines before.

Reply:
No, wait, the `\t` is supposed to go there.

Reply:
Wouldn't that mess a bit with the widget placement if it were aligned to the right side of the screen?

Reply:
Please keep the tab between leveltime and use time and remove it from the end. If you right-align the widget and draw it with the Doom font, it jumps back and forth as it is now.

Reply:
The latest master itself seems to have some jumping issues:

- Right-align the Time widget;
- Use the Doom font;
- Increase game speed a bit.

At some point, the whole widget starts jumping around.

Apart from that: using the Doom font, if I put the tab back where it was, the Time part stops jumping around actively, although it does move around as the width of the Use time varies, but the Use time itself sometimes clips into the right edge of the screen and therefore makes some characters disappear. I tried some combinations of characters between the Time and Use time, and at the end of the latter, but none of them worked properly; I'm not sure what to do.

Reply:
> At some point, the whole widget starts jumping around.

Yep, right-aligning variable-length strings is fun. To mitigate the issue, I append the tab char to the end of the string which rounds the string length up to the next multiple of 16. So, it remains stable for most string widths. But for others, a single pixel may make the difference between *this* multiple of 16 and *the next* multiple of 16.

> Apart from that: using the Doom font, if I put the tab back where it was, the Time part stops jumping around actively, although it does move around as the width of the Use time varies, 

That's how it should be.

> but the Use time itself sometimes clips into the right edge of the screen and therefore makes some characters disappear. 

This must not happen!

> I tried some combinations of characters between the Time and Use time, and at the end of the latter, but none of them worked properly; I'm not sure what to do.

Plase leave a space between them and a tab at the end. The clipping problem must be fixed elsewhere.

Reply:
> Plase leave a space between them and a tab at the end. The clipping problem must be fixed elsewhere.

Bullshit, the other way round, please. :goberserk:

Reply:
And even this is bullshit, we don't need an additional char at the end. I need more sleep...

Reply:
There's already a space between Time and Use time, and a tab at the end of the string. You want me to add a tab before the space?

It'd look something like `[Time]\t [Use time]\t`. Is that it?

Reply:
I guess it needs something like this to fix tabbed printing:

```patch
--- a/src/hu_lib.c
+++ b/src/hu_lib.c
@@ -309,6 +309,7 @@ static int vert_align_widget(const hu_widget_t *const w, const hu_multiline_t *c

 static void draw_line_aligned (const hu_multiline_t *m, const hu_line_t *l, const hu_font_t *const f, int x, int y)
 {
+  const int x0 = x;
   int i;
   unsigned char c;
   char *cr = m->cr;
@@ -328,7 +329,7 @@ static void draw_line_aligned (const hu_multiline_t *m, const hu_line_t *l, cons
 #endif
     if (c == '\t')    // killough 1/23/98 -- support tab stops
     {
-      x = (x + f->tab_width) & f->tab_mask;
+      x = x0 + (((x - x0) + f->tab_width) & f->tab_mask);
     }
     else if (c == '\x1b')  //jff 2/17/98 escape code for color change
     {               //jff 3/26/98 changed to actual escape char
```

Reply:
> I'm not sure what to do.

How about this?

```patch
--- a/src/hu_stuff.c
+++ b/src/hu_stuff.c
@@ -1170,7 +1170,7 @@ static void HU_widget_build_sttime(void)
               '0'+CR_GREEN, time/60, time%60);
     }

-    offset += sprintf(hud_timestr + offset, "\x1b%c%d:%05.2f",
+    offset += sprintf(hud_timestr + offset, "\x1b%c%d:%05.2f\t",
                       '0'+CR_GRAY, leveltime/TICRATE/60, (float)(leveltime%(60*TICRATE))/TICRATE);

     if (plr->btuse_tics)
@@ -1185,8 +1185,6 @@ static void HU_widget_build_sttime(void)
                       (float) (plr->btuse % (60 * TICRATE)) / TICRATE);
   }

-  sprintf(hud_timestr + offset, "\t");
-
   HUlib_add_string_to_cur_line(&w_sttime, hud_timestr);
 }

```

Reply:
> I'm pretty sure that's one of the combinations I tried and had issues with. I'll check again later just in case.

Yes, and please apply the fix I posted here:
https://github.com/fabiangreffrath/woof/pull/1414#discussion_r1472664330

Reply:
> please apply the fix I posted here: [#1414 (comment)](https://github.com/fabiangreffrath/woof/pull/1414#discussion_r1472664330)

In this PR? Just making sure, as it feels a bit out of scope.

Reply:
I could apply this to the master branch first and you merge the master branch into your PR branch.

Reply:
I think that'd be better, yes. Go ahead.

Reply:
Still not perfect, but better than before. Please proceed.

Reply:
I'm afraid to say that the latest master still has the jumping issues described earlier. _However,_ I applied your tab fix in b4151c07fc211109b4cef4476dd6d7f91fe18454 and your change to `HU_widget_build_sttime()` in 1cd5e30dd5dce8b36dd527aefdaee64746d92f26, and they seem to have fixed the jumping with the Use time -- at least I couldn't provoke any jumping yet.